be/src/olap/rowset/segment_v2/segment_writer.cpp

```diff
@@ -811,7 +813,18 @@ Status SegmentWriter::append_block(const vectorized::Block* block, size_t row_po
         }
         RETURN_IF_ERROR(_column_writers[id]->append(converted_result.second->get_nullmap(),
                                                     converted_result.second->get_data(), num_rows));
+
+        // estimate column data size for flush memtable, may be inaccurate at low cardinality
+        column_data_size += _column_writers[id]->estimate_buffer_size();
+        total_data_size += column_data_size;
+        auto origin_column_data_size = _footer.columns(id).total_data_size();
+        _footer.mutable_columns(id)->set_total_data_size(origin_column_data_size +
+                                                         column_data_size);
     }
+
+    auto origin_data_footprint = _footer.data_footprint();
+    _footer.set_data_footprint(origin_data_footprint + total_data_size);
```

Comment:
you don't need to accumulate these stats in each `append_block`, you can get the final stats once in the `finalize` method